### ISSUE TITLE
Proof: absence of runtime errors

### ISSUE THREAD

**Issue by ArthurClaviere** [yoogx/Certyflie#5]
Need to prove the following unproved:

commander.adb:146:48: medium: range check might fail
commander.adb:147:48: medium: range check might fail
commander.adb:148:28: medium: range check might fail
stabilizer.adb:213:36: medium: range check might fail
stabilizer.adb:213:55: medium: range check might fail
stabilizer.adb:214:36: medium: range check might fail
stabilizer.adb:260:55: medium: range check might fail
stabilizer.adb:392:33: medium: range check might fail
stabilizer.adb:393:33: medium: range check might fail
stabilizer.ads:108:08: low: global input "Stabilizer.Motor_Powers" of "Stabilizer_Control_Loop" not read

**Comment by yannickmoy** [comment]
The first 3 are indeed not provable:

commander.adb:146:48: medium: range check might fail
commander.adb:147:48: medium: range check might fail
commander.adb:148:28: medium: range check might fail

Here we call a procedure taking in-out parameters of type Float, and we pass in arguments of type T_Degrees which range from -360.0 to 360.0. So if the procedure returns a value outside these bounds for these parameters, a range check failure will occur.

As far I can see, there is the same problem for every other message: the procedure takes Float as in-out parameters, and at the call site variables of more precise types are passed as arguments.

Regarding the last message:

stabilizer.ads:108:08: low: global input "Stabilizer.Motor_Powers" of "Stabilizer_Control_Loop" not read

The tool claims that the implementation only updates Stabilizer.Motor_Powers, while it is declared as a Global In_Out. Hence the discrepancy. To be checked with the developers of Certyflie.

**Comment by ArthurClaviere** [comment]
Indeed, some VCs are not provable, that's why we modified the code by taking account of the data exchanged. 
For example, for the first 3, we changed Float to the expected type T_Degrees.

**Comment by pierreciv** [comment]
About the message : 
stabilizer.ads:108:08: low: global input "Stabilizer.Motor_Powers" of "Stabilizer_Control_Loop" not read

The issue is fixed if we change the specification like below : 

procedure Stabilizer_Control_Loop
     (Attitude_Update_Counter : in out T_Uint32;
      Alt_Hold_Update_Counter : in out T_Uint32)
     with
       Global => (Input  => (V_Speed_Parameters,
                             Asl_Parameters,
                             Alt_Hold_Parameters,
                             IMU_State,
                             Power_Management_State,
                             Clock_Time),
                  **Output => (Motor_Powers),**
                  In_Out => (Commander_State,
                             CRTP_State,
                             FF_State,
                             Motors_State,
                             SensFusion6_State,
                             IMU_Outputs,
                             Desired_Angles,
                             Desired_Rates,
                             Actual_Angles,
                             Command_Types,
                             Actuator_Commands,
                             **--  Motor_Powers,**
                             Attitude_PIDs,
                             Rate_PIDs,
                             V_Speed_Variables,
                             Asl_Variables,
                             Alt_Hold_Variables));

**However**, that solution seems strange. Indeed, The procedure Stabilizer_Control_Loop call the procedure Stabilizer_Distribute_Power with the specification below : 

procedure Stabilizer_Distribute_Power
     (Thrust : T_Uint16;
      Roll   : T_Int16;
      Pitch  : T_Int16;
      Yaw    : T_Int16)
     with
       Global => (**Output => Motor_Powers**,
                  In_Out => Motors_State);

which suggest that Motor_Powers is not read, but we have : 

procedure Stabilizer_Distribute_Power 
(Thrust : T_Uint16; Roll   : T_Int16 ;  Pitch  : T_Int16 ; Yaw    : T_Int16)
   is
   [...]       
   begin
   [...]
      Motor_Set_Power_With_Bat_Compensation (MOTOR_M1, **Motor_Power_M1**);
      Motor_Set_Power_With_Bat_Compensation (MOTOR_M2, Motor_Power_M2);
      Motor_Set_Power_With_Bat_Compensation (MOTOR_M3, Motor_Power_M3);
      Motor_Set_Power_With_Bat_Compensation (MOTOR_M4, Motor_Power_M4);
   end Stabilizer_Distribute_Power;

which suggests that Motor_Powers is read as well

Finally, if we change the specification of the procedure Stabilizer_Distribute_Power with the specification below : 

procedure Stabilizer_Distribute_Power
     (Thrust : T_Uint16;
      Roll   : T_Int16;
      Pitch  : T_Int16;
      Yaw    : T_Int16)
     with
       Global => (**In_Out => Motor_Powers**,
                                         Motors_State);

we get other law-issue telling : global input "Stabilizer.Motor_Power_M1" not read.

As a conclusion , the issue seems to have been fixed , but the reason is unclear for me.

**Comment by yannickmoy** [comment]
GNATprove is right on all cases.

Inside Stabilize_Distribute_Power, all four components Motor_Power_M? of Motor_Powers are assigned (in one branch of the if-statement) before reading it in the call you mentioned. So Motor_Powers is really an output of Stabilize_Distribute_Power.
